distributed/worker.py

```diff
@@ -157,6 +157,37 @@
 DEFAULT_STARTUP_INFORMATION: dict[str, Callable[[Worker], Any]] = {}
 
 
+def fail_hard(method):
+    """
+    Decorator to close the worker if this method encounters an exception
+    """
+    if iscoroutinefunction(method):
+
+        @functools.wraps(method)
+        async def wrapper(self, *args, **kwargs):
+            try:
+                return await method(self, *args, **kwargs)
+            except Exception as e:
```

Comment:
I'm OK with catching BaseException for tasks, i.e. in `apply_function`, et al. to fix #5958

However, I would be worried to close workers upon a `asyncio.CancelledError`. While I don't think we're using cancellation in many places right now, this would be a very confusing behavior if that ever changes.